crates/router/src/core/payments.rs

```diff
@@ -1852,6 +1818,85 @@ where
     Ok((router_data, merchant_connector_account))
 }
 
+pub async fn add_decrypted_payment_method_token<F, Req, D>(
+    tokenization_action: TokenizationAction,
+    payment_data: &D,
+    mut router_data: RouterData<F, Req, router_types::PaymentsResponseData>,
+) -> RouterResult<RouterData<F, Req, router_types::PaymentsResponseData>>
+where
+    F: Send + Clone + Sync,
+    D: OperationSessionGetters<F> + Send + Sync + Clone,
+    Req: Send + Sync,
+{
+    // Tokenization Action will be DecryptApplePayToken, only when payment method type is Apple Pay
+    // and the connector supports Apple Pay predecrypt
+    match &tokenization_action {
+        TokenizationAction::DecryptApplePayToken(payment_processing_details)
+        | TokenizationAction::TokenizeInConnectorAndApplepayPreDecrypt(
+            payment_processing_details,
+        ) => {
+            let apple_pay_data = match payment_data.get_payment_method_data() {
+                Some(domain::PaymentMethodData::Wallet(domain::WalletData::ApplePay(
+                    wallet_data,
+                ))) => Some(
+                    ApplePayData::token_json(domain::WalletData::ApplePay(wallet_data.clone()))
+                        .change_context(errors::ApiErrorResponse::InternalServerError)?
```

Comment:
Can we add context messages with `attach_printable()` / `attach_printable_lazy()` wherever we've raised internal server errors?

Reply:
Done